src/tagsouppullparser.cpp

```diff
@@ -149,23 +149,25 @@ std::string TagSoupPullParser::decode_attribute(const std::string& s)
 	return decode_entities(s1);
 }
 
-std::string TagSoupPullParser::decode_entities(const std::string& s)
+std::string TagSoupPullParser::decode_entities(std::string s)
 {
-	std::string result;
-	std::istringstream sbuf(s);
-	std::string tmp;
-	getline(sbuf, tmp, '&');
-	while (!sbuf.eof()) {
-		result.append(tmp);
-		getline(sbuf, tmp, ';');
-		if (sbuf.eof()) {
-			break;
+	size_t offset = 0;
+	size_t semicolon_offset;
+	std::string encoded_entity;
+	while ((offset = s.find('&', offset)) != std::string::npos) {
+		offset++;
+		semicolon_offset = s.find(';', offset);
+		if (semicolon_offset == std::string::npos) {
+			continue;
+		}
+		encoded_entity = decode_entity(s.substr(offset, semicolon_offset - offset));
+		if (encoded_entity.empty()) {
+			continue;
 		}
-		result.append(decode_entity(tmp));
-		getline(sbuf, tmp, '&');
+		s.replace(offset - 1, semicolon_offset - offset + 2, encoded_entity);
```

Comment:
`std::string::replace` is copies the "tail" after the given position into the new position, i.e. it has linear time complexity. Using it in a loop makes the entire algorithm quadratic. I don't have benchmarks to prove that this is a severe issue, but the issue seems easy enough to fix: just keep the result in a separate string, and append to it (like the previous algorithm did).

(Don't forget to undo the change to the method signature, since there would be no point in making a copy of the argument then.)